src/maliput_malidrive/road_curve/spiral_ground_curve.cc

```diff
@@ -0,0 +1,200 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#include "maliput_malidrive/road_curve/spiral_ground_curve.h"
+
+#include <algorithm>
+#include <cmath>
+
+#include <maliput/math/fresnel.h>
+
+namespace malidrive {
+namespace road_curve {
+namespace {
+
+// Rotates @p v by @p angle.
+// @param v Input 2D vector to rotate.
+// @param angle Input angle to rotate.
+// @return The rotated vector.
+inline maliput::math::Vector2 RotateVector(const maliput::math::Vector2& v, double angle) {
+  const double ca = std::cos(angle);
+  const double sa = std::sin(angle);
+  return maliput::math::Vector2{ca * v.x() - sa * v.y(), sa * v.x() + ca * v.y()};
+}
+
+// Conditionally shifts the y-component of @p v when @p cond is true.
+// @param v Input vector.
+// @param cond Conditional flag.
+// @return When @p cond is true, returns `[v.x(), -v.y()]`, otherwise @p v.
+inline maliput::math::Vector2 ConditionalShiftYComponent(const maliput::math::Vector2& v, bool cond) {
+  return cond ? maliput::math::Vector2{v.x(), -v.y()} : v;
+}
+
+// Finds the best p parameter for @p curve that maps a point in the 2D INERTIAL-Frame that minimizes
+// the distance to @p point.
+// @details The function implements a binary search within a range trying to minimize the distance against
+// @p point. The range is split into equal @p partitions. The search can be early terminated if any of the
+// extents of the range at each iteration is within @p tolerance.
+// @param curve The SpiralGoundCurve to use.
+// @param point The point in the 2D INERTIAL-Frame to find a p parameter value for.
+// @param p0 The p parameter start range value. It must be non-negative.
+// @param p1 The p parameter end range value. It must be greater than @p p0 by GroundCurve::kEpsilon.
+// @param partitions The number of partitions of the range [@p p0; @p p1]. It must be greater or equal to 2.
+// @param tolerance The tolereance to early stop the search. It must be non-negative.
+// @return The p parameter value that minimizes the distance against @p point.
+double FindBestPParameter(const SpiralGroundCurve& curve, const maliput::math::Vector2& point, double p0, double p1,
+                          size_t partitions, double tolerance) {
```

Comment:
If we notice this implementation does not perform well we could use the kdtree to boost this up.

Reply:
Ack!